### ISSUE TITLE
Webmin/Virtualmin uses the root password method and overwrites mariadb's default authentication methods, resulting in broken OS database upgrades routines

### ISSUE THREAD

**Issue by nekohayo** [virtualmin/virtualmin-gpl#648]
## TLDR summary

This bug report is two parts: the current problem, and the alternative approach / futureproof solution.

* Part 1, "**current problem**": Webmin 2.101 and Virtualmin 7.7 / 7.8.2 set the mariadb root password in a way that accidentally removes the `unix_socket` authentication method, which is used by systemd and various distro upgrade scripts.
  * As part of the initial [installation/setup script](https://software.virtualmin.com/gpl/scripts/virtualmin-install.sh), with code like [feature-mysql.pl](https://github.com/virtualmin/virtualmin-gpl/blob/HEAD/feature-mysql.pl) doing things like  
  `$sql = "alter user '$user'\@'$host' identified $plugin by '".&mysql_escape($plainpass)."'";`  
  within `sub execute_password_change_sql` ;
  * The problem with that is, when you just `alter user root identified by password("whatever")`, you throw away the `unix_socket` auth method... and your users won't notice until they do a major version upgrade of MariaDB that requires a database upgrade (with `mariadb-upgrade`) and the operating system fails to do the automatic upgrade because the passwordless `unix_socket` method is unavailable. See the investigation details below for more details on how I came to this conclusion.

* Part 2, "**additional futureproof / cleaner solution for Webmin/Virtualmin itself**": since MariadDB 10.4 and newer (so, Debian 11 and newer), the root password method is discouraged by mariadb. It would be smarter for webmin and virtualmin to not need this at all, and just use the unix socket method to control mariadb as a superuser. See [this page from the MariaDB docs](https://mariadb.com/kb/en/authentication-from-mariadb-10-4/), which says, among other things:

> As a result of the above changes, the open-for-everyone all-powerful root account is finally gone. And installation scripts will no longer demand that you “PLEASE REMEMBER TO SET A PASSWORD FOR THE MariaDB root USER !”, because the root account is securely created automatically.
> [...]
> It is based on a simple fact that asking the system root for a password adds no extra security — root has full access to all the data files and all process memory anyway. But not asking for a password means, there is no root password to forget (no need for the numerous tutorials on “how to reset MariaDB root password”). And if you want to script some tedious database work, there is no need to store the root password in plain text for the script to use (no need for debian-sys-maint user).
> [...]
> Even if MariaDB is installed system-wide, you may not want to run your database maintenance scripts as system root — now you can run them as system mysql user. And you will know that they will never destroy your entire system, even if you make a typo in a shell script.

Thank you for coming to my TED talk 👏

---

## The rabbithole's context / how this was a problem for me in practice

I was wondering why my "Debian 11 + Virtualmin 7.7 ugraded to Debian 12 + Virtualmin 7.8.2" server was constantly spamming my system logs with these errors:

```
[ERROR] Incorrect definition of table mysql.column_stats:
expected column 'hist_type' at position 9 to have type enum('SINGLE_PREC_HB','DOUBLE_PREC_HB','JSON_HB'),
found type enum('SINGLE_PREC_HB','DOUBLE_PREC_HB').
```

...and it turns out that to fix this, I simply had to do  `mariadb-upgrade -p`. I mentioned this to @s3phy, and he wondered why I even had to do run that upgrade command manually (we would have expected debian's post-install package upgrade scripts to take care of this, regardless of the state of the configuration files), so below is the rabbithole we went into.

On his server, without webmin/virtualmin:

```
$ sudo mariadb-upgrade -u root -p
This installation of MariaDB is already upgraded to 10.11.3-MariaDB.
There is no need to run mysql_upgrade again for 10.11.3-MariaDB.
You can use --force if you still want to run mysql_upgrade
```

...and he never had to run that command before, so it did work automatically, because Debian's package has provisions for this:

* https://sources.debian.org/src/mariadb/1%3A10.11.3-1/debian/mariadb-server.postinst/
* https://sources.debian.org/src/mariadb/1%3A10.11.3-1/debian/additions/debian-start/

In fact, with systemd, this stuff is supposed to happen whenever the mariadb service starts, even! `systemctl status mariadb.service`:

```
Loaded: loaded (/lib/systemd/system/mariadb.service; enabled; preset: enabled)
Active: active (running) since Fri 2023-09-22 11:26:11 EDT; 1h 32min ago
```

...then looking in `/usr/lib/systemd/system/mariadb.service` we have a line that says `ExecStartPost=/etc/mysql/debian-start`

So something specific to webmin/virtualmin servers is blocking upgrade routines on Debian servers (and probably other modern OSes).

So now we look at `/etc/mysql/debian.cnf` and we see this on my server:

```
# THIS FILE IS OBSOLETE. STOP USING IT IF POSSIBLE.
# This file exists only for backwards compatibility for
# tools that run '--defaults-file=/etc/mysql/debian.cnf'
# and have root level access to the local filesystem.
# With those permissions one can run 'mariadb' directly
# anyway thanks to unix socket authentication and hence
# this file is useless. See package README for more info.
[client]
host     = localhost
user     = root
[mysql_upgrade]
host     = localhost
user     = root
# THIS FILE WILL BE REMOVED IN A FUTURE DEBIAN RELEASE.
```

In <https://sources.debian.org/src/mariadb/1%3A10.11.3-1/debian/mariadb-server.postinst/> we see this further explanation comment:

> On new installations root user can connect via unix_socket.  
> But on upgrades, scripts rely on debian-sys-maint user and credentials in /etc/mysql/debian.cnf  
> All tools use `--defaults-file=/etc/mysql/debian.cnf`
> And while it's not needed for new installations, we keep using --defaults-file option for tools (for the sake of upgrades) and thus need /etc/mysql/debian.cnf to exist, even if it's empty.  
> In the long run the goal is to obsolete this file.

Now, if I do `mariadb -u root` as root or via sudo, I get prompted for a password, whether I use the `-p` parameter or not... but I shouldn't, because as mentioned in the beginning of this bug report, [MariadDB 10.4+ doesn't require specifying a password when the user already has root privileges](https://mariadb.com/kb/en/authentication-from-mariadb-10-4/). And yet, on my system:

```
ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: NO)
```

...which explains why the upgrade routines didn't work, because Debian depends on those passwordless tricks to work, and virtualmin/webmin removed root's privileges to mariadb.

The `SHOW GRANTS;` SQL query lets you see the state of things.  
On a non-webmin/non-virtualmin server, it is:

```
GRANT ALL PRIVILEGES ON *.* TO `root`@`localhost` IDENTIFIED VIA unix_socket OR mysql_native_password USING '<some_big_hash>' WITH GRANT OPTION
GRANT PROXY ON ''@'%' TO 'root'@'localhost' WITH GRANT OPTION
```

On my virtualmin server, it is:

```
GRANT ALL PRIVILEGES ON *.* TO `root`@`localhost` IDENTIFIED BY PASSWORD '*some_big_hash' WITH GRANT OPTION
GRANT PROXY ON ''@'%' TO 'root'@'localhost' WITH GRANT OPTION
```

The only way to revert/fix this as a workaround for now would be this SQL query (and obviously, my password is not "1234" ;) :

```
ALTER USER root@localhost IDENTIFIED VIA unix_socket OR mysql_native_password USING PASSWORD("1234")
```

...but this would be overwritten again whenever webmin/virtualmin's code (such as [feature-mysql.pl](https://github.com/virtualmin/virtualmin-gpl/blob/HEAD/feature-mysql.pl)) decides to update the root password by using `alter user` and blindly overwriting things without preserving the `unix_socket` method.

I [searched](https://google.com/search?hl=en&q=webmin%20OR%20virtualmin%20mariadb%20%22unix_socket%22%20site%3Agithub.com%20OR%20site%3Aforum.virtualmin.com) for issues related to this and I saw nothing relevant, so I guess I'm the first one to run into it ahead of the curve.

---

Presumably webmin/virtualmin already run with system / root privileges, so it would be better for them to also use the passwordless `unix_socket` method for themselves on modern versions of mariadb, instead of depending on setting and using a mariadb root password.

_Potentially_ also related (?) : https://github.com/virtualmin/virtualmin-gpl/issues/646

Additional docs if needed: https://mariadb.com/kb/en/pluggable-authentication-overview/

**Comment by jcameron** [comment]
Seems like the right thing would be for Webmin / Virtualmin to not change the root password if the `unix_socket` method is enabled AND we're connecting to local MySQL, on the assumption that it doesn't need a password to connect anyway.

**Comment by xxorax** [comment]
I went deep inside this problem some times ago.

The main issue start when [debian changed his authentification method](https://salsa.debian.org/mariadb-team/mariadb-10.5/-/commit/6937a348609cdb748e3c1e627f81b88bb6bf60b0#9a821da56fc37603af138f456945d21a6d00a5e7_139_142) for maintenance task (like upgrade) : they no longer use the debian-sys-maint user and rely now on the root user with unix_socket.

However, using only using_socket means you cannot connect to mysql root except by connecting to linux root user, and using this socket. Means that you need to share root access and your tools have to handle unix socket (though ssh for example) : it's sad.

On my side I have created a post install command to re-add the debian-sys-maint user, as it was before, using password auth :

```bash
user=debian-sys-maint
host=localhost
pw="$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 16 | head -n 1)"
f=/etc/mysql/debian.cnf

mysql -e "GRANT ALL ON *.* TO '$user'@'$host' IDENTIFIED BY '${pw}' WITH GRANT OPTION;"

sed -i -r --follow-symlinks \
    -e 's/^( *user *=).+$/\1 '$user'/' \
    -e 's/^( *password *=).+$/\1 '$pw'/' \
	$f

# if the password line was missing
if ! grep -q password $f ; then
	sed -i -r  --follow-symlinks '/^user *=.+$/a password = '"$pw" $f
fi
```

It will work until Debian will remove /etc/mysql/debian.cnf as it is marked as obsolete...

**Comment by jcameron** [comment]
Do you know which Debian version they introduced this change in?

**Comment by xxorax** [comment]
> Do you know which Debian version they introduced this change in?

The [changelog](https://metadata.ftp-master.debian.org/changelogs//main/m/mariadb/mariadb_10.11.5-1_changelog) mention it in 10.0 ( mariadb-10.0 (10.0.16-1~exp3) : "Stop asking and setting a database root user password...")
But I can  find the patch only in [Mariadb-10.1](https://salsa.debian.org/mariadb-team/mariadb-10.1/-/commit/a4b4c95a5cb4e0ce053e8e3d04608ec0a2b03fbc)

I finally think you are right in your [first comment](https://github.com/virtualmin/virtualmin-gpl/issues/648#issuecomment-1732169430). Notice that in older version unix_socket was auth_socket.

Also, since Mariabd 10.4 you can use [multiple auth pluigin](https://mariadb.com/kb/en/authentication-from-mariadb-10-4/ ) `IDENTIFIED VIA unix_socket OR mysql_native_password`(not sure we should bring an other level of version specific complexity...)

And at the end if anyone (like me) want to allow password auth, we have plenty of choices (use multiple auth plugin on the root account, or add an other user with root privileges)

**Comment by jcameron** [comment]
What I'll do for starters is prevent the `root` password from being changed in the MySQL module when socket authentication is in use..

**Comment by iliajie** [comment]
Jamie, would you like to keep this ticket opened or have it closed?

**Comment by jcameron** [comment]
No I think this is fixed now
